### PR TITLE
Add snapshot secret reference to group snapshot controller.

### LINKED ISSUE #834: Add secrets in CreateVolumeGroupSnapshots

### PR DESCRIPTION
**Pull request by manishym**
Fixes: https://github.com/kubernetes-csi/external-snapshotter/issues/834

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

 /kind feature
> /kind flake

**What this PR does / why we need it**: 
Adds secret reference to group snapshots

**Which issue(s) this PR fixes**:  
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #834

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Adds secret reference to volume group snapshot content.
```

### CODE REVIEW

File: pkg/common-controller/groupsnapshot_controller_helper.go [hunk 1]
```diff
@@ -819,23 +826,31 @@ func (ctrl *csiSnapshotCommonController) getCreateGroupSnapshotInput(groupSnapsh
...
 	}
 
 	// Create VolumeGroupSnapshotContent name
 	contentName := utils.GetDynamicSnapshotContentNameForGroupSnapshot(groupSnapshot)
 
-	return groupSnapshotClass, volumes, contentName, nil
+	// Get the secret reference
+	snapshotterSecretRef, err := utils.GetGroupSnapshotSecretReference(utils.SnapshotterSecretParams, groupSnapshotClass.Parameters, contentName, groupSnapshot)
+
+	klog.V(5).Infof("snapshotterSecretRef %+v", snapshotterSecretRef)
```
Comment (Madhu-1) [inline pkg/common-controller/groupsnapshot_controller_helper.go:848]: can we remove this log?

File: pkg/sidecar-controller/groupsnapshot_helper.go [hunk 2]
```diff
@@ -423,6 +423,11 @@ func (ctrl *csiSnapshotSideCarController) createGroupSnapshotWrapper(groupSnapsh
 		creationTime = time.Now()
 	}
 
+	groupSnapshotSecret, err := utils.GetSecretReference(utils.SnapshotterSecretParams, class.Parameters, groupSnapshotContent.GetObjectMeta().GetName(), nil)
```
Comment (Madhu-1) [inline pkg/sidecar-controller/groupsnapshot_helper.go:426]: instead of `utils.SnapshotterSecretParams` can use have group specific secrets? like below

```go
PrefixedGroupSnapshotterSecretNameKey   = csiParameterPrefix + "group-snapshotter-secret-name"      // Prefixed name key for CreateGroupSnapshot secret
PrefixedGroupSnapshotterSecretNamespaceKey = csiParameterPrefix + "group-snapshotter-secret-namespace" // Prefixed namespace key for DeleteGroupSnapshot secret
var GroupSnapshotterSecretParams = secretParamsMap{
	name:               "Snapshotter",
	secretNameKey:      PrefixedGroupSnapshotterSecretNameKey,
	secretNamespaceKey: PrefixedGroupSnapshotterSecretNamespaceKey,
}
```

File: pkg/sidecar-controller/groupsnapshot_helper.go [hunk 3]
```diff
@@ -498,7 +510,7 @@ func (ctrl *csiSnapshotSideCarController) createGroupSnapshotWrapper(groupSnapsh
 
 func (ctrl *csiSnapshotSideCarController) getCSIGroupSnapshotInput(groupSnapshotContent *crdv1alpha1.VolumeGroupSnapshotContent) (*crdv1alpha1.VolumeGroupSnapshotClass, map[string]string, error) {
 	className := groupSnapshotContent.Spec.VolumeGroupSnapshotClassName
-	klog.V(5).Infof("getCSIGroupSnapshotInput for group snapshot content [%s]", groupSnapshotContent.Name)
+	klog.V(5).Infof("getCSIGroupSnapshotInput for group snapshot content [%+v]", groupSnapshotContent)
```
Comment (Madhu-1) [inline pkg/sidecar-controller/groupsnapshot_helper.go:513]: revert this change?

File: pkg/sidecar-controller/snapshot_controller.go [hunk 4]
```diff
@@ -265,7 +266,7 @@ func (ctrl *csiSnapshotSideCarController) checkandUpdateContentStatusOperation(c
 				return content, fmt.Errorf("failed to get snapshot class %s for snapshot content %s: %v", *content.Spec.VolumeSnapshotClassName, content.Name, err)
 			}
 
-			snapshotterListSecretRef, err := utils.GetSecretReference(utils.SnapshotterListSecretParams, class.Parameters, content.GetObjectMeta().GetName(), nil)
+			snapshotterListSecretRef, err := utils.GetSecretReference(utils.SnapshotterSecretParams, class.Parameters, content.GetObjectMeta().GetName(), nil)
```
Comment (Madhu-1) [inline pkg/sidecar-controller/snapshot_controller.go:269]: This also need to be reverted?

File: pkg/sidecar-controller/snapshot_controller.go [hunk 5]
```diff
@@ -560,6 +561,36 @@ func (ctrl *csiSnapshotSideCarController) GetCredentialsFromAnnotation(content *
 	return snapshotterCredentials, nil
 }
 
+func (ctrl *csiSnapshotSideCarController) GetGroupCredentialsFromAnnotation(content *crdv1alpha1.VolumeGroupSnapshotContent) (map[string]string, error) {
```
Comment (Madhu-1) [inline pkg/sidecar-controller/snapshot_controller.go:564]: Move this to groupsnapshtot_helper?

File: pkg/sidecar-controller/snapshot_controller.go [hunk 6]
```diff
@@ -560,6 +561,36 @@ func (ctrl *csiSnapshotSideCarController) GetCredentialsFromAnnotation(content *
...
+	// get secrets if VolumeSnapshotClass specifies it
```
Comment (Madhu-1) [inline pkg/sidecar-controller/snapshot_controller.go:565]: we are not getting the secret from class, this comment can be removed.

File: pkg/sidecar-controller/snapshot_controller.go [hunk 7]
```diff
@@ -560,6 +561,36 @@ func (ctrl *csiSnapshotSideCarController) GetCredentialsFromAnnotation(content *
...
+	var snapshotterCredentials map[string]string
+	var err error
+
+	// Check if annotation exists
+	if metav1.HasAnnotation(content.ObjectMeta, utils.AnnDeletionSecretRefName) && metav1.HasAnnotation(content.ObjectMeta, utils.AnnDeletionSecretRefNamespace) {
+		annDeletionSecretName := content.Annotations[utils.AnnDeletionSecretRefName]
+		annDeletionSecretNamespace := content.Annotations[utils.AnnDeletionSecretRefNamespace]
+
+		snapshotterSecretRef := &v1.SecretReference{}
```
Comment (Madhu-1) [inline pkg/sidecar-controller/snapshot_controller.go:574]: instead of `snapshotterSecret` use group specific names in other places in this function

File: pkg/sidecar-controller/snapshot_controller.go [hunk 8]
```diff
@@ -560,6 +561,36 @@ func (ctrl *csiSnapshotSideCarController) GetCredentialsFromAnnotation(content *
...
+
+		if annDeletionSecretName == "" || annDeletionSecretNamespace == "" {
+			return nil, fmt.Errorf("cannot retrieve secrets for snapshot content %#v, err: secret name or namespace not specified", content.Name)
```
Comment (Madhu-1) [inline pkg/sidecar-controller/snapshot_controller.go:577]:
```suggestion
return nil, fmt.Errorf("cannot retrieve secrets for volume group snapshot content %v, err: secret name or namespace not specified", content.Name)
```

File: pkg/utils/util.go [hunk 9]
```diff
@@ -284,6 +284,7 @@ func verifyAndGetSecretNameAndNamespaceTemplate(secret secretParamsMap, snapshot
 		numNamespace++
 	}
 
+	klog.V(5).Infof("numName %v nameTemplate %v, namespaceTemplate %v", numName, nameTemplate, namespaceTemplate)
```
Comment (Madhu-1) [inline pkg/utils/util.go:287]: remove this log

File: pkg/utils/util.go [hunk 10]
```diff
@@ -376,6 +377,64 @@ func GetSecretReference(secretParams secretParamsMap, snapshotClassParams map[st
 	return ref, nil
 }
 
+// GetSecretReference for the group snapshot
+func GetGroupSnapshotSecretReference(secretParams secretParamsMap, snapshotClassParams map[string]string, snapContentName string, snapshot *crdv1alpha1.VolumeGroupSnapshot) (*v1.SecretReference, error) {
```
Comment (Madhu-1) [inline pkg/utils/util.go:381]: `snapshotClassParams` to `volumeGroupSnapshotClassParams` , `snapContentName` to `groupSnapshotContentName` and `snapshot` to `volumeGroupSnapshot`

File: pkg/utils/util.go [hunk 11]
```diff
@@ -376,6 +377,64 @@ func GetSecretReference(secretParams secretParamsMap, snapshotClassParams map[st
...
+	// snapshot may or may not exist at delete time
+	if snapshot != nil {
+		namespaceParams["volumegroupsnapshot.namespace"] = snapshot.Namespace
+	}
+
+	resolvedNamespace, err := resolveTemplate(namespaceTemplate, namespaceParams)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving value %q: %v", namespaceTemplate, err)
+	}
+	klog.V(4).Infof("GetGroupSnapshotSecretReference namespaceTemplate %s, namespaceParams: %+v, resolved %s", namespaceTemplate, namespaceParams, resolvedNamespace)
```
Comment (Madhu-1) [inline pkg/utils/util.go:405]: Remove this log as well

File: pkg/utils/util.go [hunk 12]
```diff
@@ -376,6 +377,64 @@ func GetSecretReference(secretParams secretParamsMap, snapshotClassParams map[st
...
+	}
+	if len(validation.IsDNS1123Subdomain(resolvedName)) > 0 {
+		if nameTemplate != resolvedName {
+			return nil, fmt.Errorf("%q resolved to %q which is not a valid secret name", nameTemplate, resolvedName)
+		}
+		return nil, fmt.Errorf("%q is not a valid secret name", nameTemplate)
+	}
+	ref.Name = resolvedName
+
+	klog.V(4).Infof("GetGroupSnapshotSecretReference validated Secret: %+v", ref)
```
Comment (Madhu-1) [inline pkg/utils/util.go:434]: Remove this log

File: pkg/sidecar-controller/groupsnapshot_helper.go (hunk 2 above)
Comment (nixpanic) [reply pkg/sidecar-controller/groupsnapshot_helper.go:426]: I am not sure that there is much  value in having separate GroupSnapshot secrets. @RaunakShah or @xing-yang what do you think?

File: pkg/common-controller/groupsnapshot_controller_helper.go [hunk 13]
```diff
@@ -819,23 +826,30 @@ func (ctrl *csiSnapshotCommonController) getCreateGroupSnapshotInput(groupSnapsh
...
+		return nil, nil, "", nil, err
 	}
 
 	// Create VolumeGroupSnapshotContent name
 	contentName := utils.GetDynamicSnapshotContentNameForGroupSnapshot(groupSnapshot)
 
-	return groupSnapshotClass, volumes, contentName, nil
+	// Get the secret reference
+	snapshotterSecretRef, err := utils.GetGroupSnapshotSecretReference(utils.SnapshotterSecretParams, groupSnapshotClass.Parameters, contentName, groupSnapshot)
+
```
Comment (Madhu-1) [inline pkg/common-controller/groupsnapshot_controller_helper.go:847]: No need for a new line here.

File: pkg/utils/util.go (hunk 9 above)
Comment (Madhu-1) [reply pkg/utils/util.go:296]: Still need to be removed

File: pkg/utils/util.go (hunk 11 above)
Comment (Madhu-1) [reply pkg/utils/util.go:414]: same here

File: pkg/sidecar-controller/snapshot_controller.go (hunk 5 above)
Comment (Madhu-1) [reply pkg/sidecar-controller/snapshot_controller.go:564]: still need to addressed.

**Review by Madhu-1 — COMMENTED** [review]
small nits, overall LGTM

File: pkg/utils/util.go [hunk 14]
```diff
@@ -150,6 +153,12 @@ var SnapshotterSecretParams = secretParamsMap{
 	secretNamespaceKey: PrefixedSnapshotterSecretNamespaceKey,
 }
 
+var GroupSnapshotterSecretParams = secretParamsMap{
+	name:               "Snapshotter",
```
Comment (Madhu-1) [inline pkg/utils/util.go:157]: Name should `GroupSnapshotter`  as snapshotter already sed for `SnapshotterSecretParams`

File: pkg/utils/util.go [hunk 15]
```diff
@@ -376,6 +385,61 @@ func GetSecretReference(secretParams secretParamsMap, snapshotClassParams map[st
...
+		return nil, nil
+	}
+
+	ref := &v1.SecretReference{}
+
+	// Secret namespace template can make use of the VolumeGroupSnapshotContent name, VolumeGroupSnapshot name or namespace.
+	// Note that neither of those things are under the control of the VolumeGroupSnapshot user.
+	namespaceParams := map[string]string{"volumegroupsnapshotcontent.name": groupSnapContentName}
+	// snapshot may be nil when resolving create/delete volumegroupsnapshot secret names because the
+	// snapshot may or may not exist at delete time
```
Comment (Madhu-1) [inline pkg/utils/util.go:404]:
```suggestion
// volume group snapshot may be nil when resolving create/delete volumegroupsnapshot secret names because the
	// volume group snapshot may or may not exist at delete time
```

**Review by Madhu-1 — COMMENTED** [review]
small nit, LGTM

**Comment by Madhu-1** [comment]
@xing-yang @RaunakShah can you please review this PR?

File: pkg/sidecar-controller/groupsnapshot_helper.go [hunk 16]
```diff
@@ -452,6 +457,13 @@ func (ctrl *csiSnapshotSideCarController) createGroupSnapshotWrapper(groupSnapsh
 			},
 		}
 
+		if groupSnapshotSecret != nil {
+			klog.V(5).Infof("createGroupSnapshotContent: set annotation [%s] on content [%s].", utils.AnnDeletionSecretRefName, volumeSnapshotContent.Name)
```
Comment (xing-yang) [inline pkg/sidecar-controller/groupsnapshot_helper.go:461]: Can you clarify in the log that this is "volume snapshot content"?

File: pkg/sidecar-controller/groupsnapshot_helper.go [hunk 17]
```diff
@@ -452,6 +457,13 @@ func (ctrl *csiSnapshotSideCarController) createGroupSnapshotWrapper(groupSnapsh
...
+			metav1.SetMetaDataAnnotation(&volumeSnapshotContent.ObjectMeta, utils.AnnDeletionSecretRefName, groupSnapshotSecret.Name)
+
+			klog.V(5).Infof("createGroupSnapshotContent: set annotation [%s] on content [%s].", utils.AnnDeletionSecretRefNamespace, volumeSnapshotContent.Name)
```
Comment (xing-yang) [inline pkg/sidecar-controller/groupsnapshot_helper.go:464]: Can you clarify in the log that this is "volume snapshot content"?

File: pkg/common-controller/groupsnapshot_controller_helper.go [hunk 18]
```diff
@@ -773,8 +773,15 @@ func (ctrl *csiSnapshotCommonController) createGroupSnapshotContent(groupSnapsho
 	}
 
 	/*
-		TODO: Add secret reference details
+		Add secret reference details
 	*/
+	if snapshotterSecretRef != nil {
+		klog.V(5).Infof("createGroupSnapshotContent: set annotation [%s] on content [%s].", utils.AnnDeletionSecretRefName, groupSnapshotContent.Name)
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_controller_helper.go:779]: Can you clarify in the log that this is "volume group snapshot content" to differentiate from "volume snapshot content" later?

File: pkg/common-controller/groupsnapshot_controller_helper.go [hunk 19]
```diff
@@ -773,8 +773,15 @@ func (ctrl *csiSnapshotCommonController) createGroupSnapshotContent(groupSnapsho
...
+		metav1.SetMetaDataAnnotation(&groupSnapshotContent.ObjectMeta, utils.AnnDeletionSecretRefName, snapshotterSecretRef.Name)
+
+		klog.V(5).Infof("creategroupSnapshotContent: set annotation [%s] on content [%s].", utils.AnnDeletionSecretRefNamespace, groupSnapshotContent.Name)
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_controller_helper.go:782]: Same comment here.

**Comment by xing-yang** [comment]
Please add a release note.
